sdk/cosmos/azure-cosmos/azure/cosmos/_read_many_items_helper.py

```
@@ -0,0 +1,118 @@
+
+    def _partition_items_by_range(self) -> Dict[str, List[Tuple[str, "PartitionKeyType"]]]:
+        """Groups items by their partition key range ID efficiently."""
+        collection_rid = _base.GetResourceIdOrFullNameFromLink(self.collection_link)
+        partition_key = _get_partition_key_from_partition_key_definition(self.partition_key_definition)
+        items_by_partition: Dict[str, List[Tuple[str, "PartitionKeyType"]]] = {}
+
+        # Group items by logical partition key first to avoid redundant range lookups
+        items_by_pk_value: Dict[Any, List[Tuple[str, "PartitionKeyType"]]] = {}
+        for item_id, pk_value in self.items:
+            # Convert list to tuple to use as a dictionary key, as lists are unhashable
+            key = tuple(pk_value) if isinstance(pk_value, list) else pk_value
+            if key not in items_by_pk_value:
+                items_by_pk_value[key] = []
+            items_by_pk_value[key].append((item_id, pk_value))
+
+        # Now, resolve the range ID once per unique logical partition key
+        for _, pk_items in items_by_pk_value.items():
+            # All items in this list share the same partition key value. Get it from the first item.
+            pk_value = pk_items[0][1]
+            try:
+                epk_range = partition_key._get_epk_range_for_partition_key(pk_value)
+                overlapping_ranges = self.client._routing_map_provider.get_overlapping_ranges(
+                    collection_rid, [epk_range]
+                )
+                if overlapping_ranges:
+                    range_id = overlapping_ranges[0]["id"]
+                    if range_id not in items_by_partition:
+                        items_by_partition[range_id] = []
+                    items_by_partition[range_id].extend(pk_items)
+            except Exception as e:
+                self.logger.warning(f"Failed to resolve partition key range for PK '{pk_value}': {e}")
```

Would it be better to just fail here instead of returning partial results?